### PR TITLE
Initial implementation of "Pathways" plugin for Open edX

### PR DESCRIPTION
**Pull request by bradenmacdonald**
This implements LabXchange pathways as a django app plugin for Open edX.

Pathways become a new type of learning context, which has different properties and API than the existing learning contexts (libraries, courses).

For development purposes, pathways can be viewed and edited via the django admin:
http://localhost:18010/admin/lx_pathway_plugin/pathway/

Pathways are identifier via an opaque key like `lx-pathway:3f3314f8-5f59-4215-b284-09fb578f561e`

XBlocks within a pathway have their own usage keys, like `lx-pb:3f3314f8-5f59-4215-b284-09fb578f561e:unit:ec1d6c9f`.

Data about a pathway can be retrieved from the Pathway REST API, which is part of Studio - e.g. http://localhost:18010/api/lx-pathways/v1/pathway/lx-pathway:3f3314f8-5f59-4215-b284-09fb578f561e/
This provides data about both the draft and published versions of the pathway.

When content is added to a pathway, it will be frozen at that version, unless one specified `version: null` using the API, in which case it will always use the newest available version of that XBlock.

TODO:
- [x] finish REST API (update, publish/revert, delete)
- [x] permissions
- [x] write tests

### Test Instructions

Really this needs to be tested via the LabXchange app, but one can also run its edx-platform integrated test suite by opening `make studio-shell` then running:

```
make -f /edx/src/lx-pathway-plugin/Makefile validate
```

### CODE REVIEW

File: lx_pathway_plugin/views.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        pathway.save()
+        return Response(PathwaySerializer(pathway).data)
+
+    @authorized_users_only
+    def delete(self, request, pathway_key_str):
+        """
+        Delete a pathway
+        """
+        pathway = get_pathway_or_404(pathway_key_str)
+        pathway.draft_data = pathway.published_data
```
Comment (symbolist) [inline lx_pathway_plugin/views.py:115]: Hmm, what is the goal here?

Comment (bradenmacdonald) [reply lx_pathway_plugin/views.py:115]: Whoops, I need to fix the docstring. This is "revert to published version", not "delete pathway"

Comment (bradenmacdonald) [reply lx_pathway_plugin/views.py:115]: Docstring fixed.

File: lx_pathway_plugin/pathway_context.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        ID of the child.
+
+        The child is always from an <xblock-include /> element.
+        """
+        assert isinstance(parent_usage, PathwayUsageLocator)
+        # We need some kind of 'child_usage_id' part that can go into this child's
+        # usage key that provides enough information for definition_for_usage to
+        # use to get the BundleDefinitionLocator for the child (or its child...)
+        if parsed_include.link_id:
+            raise NotImplementedError("LabXchange pathways don't yet support linked content in other bundles")
```
Comment (symbolist) [inline lx_pathway_plugin/pathway_context.py:120]: Hmm. How is this going to work for assignments where a shared child problem can be added in?

**Review by symbolist — COMMENTED** [review]
I have one question but otherwise this looks good to me.

File: lx_pathway_plugin/pathway_context.py (hunk 2 above)
Comment (bradenmacdonald) [reply lx_pathway_plugin/pathway_context.py:120]: If we copy the assessment children into the assignment, then it's no problem... If we want to _link_ the assessment children into the assignment XBlock, then we'd need to (A) add an API to Open edX for creating such links, and (B) implement this code here along the lines of the comment below. We can do either one but I'm not sure we have time for the link option before launch, even though it's better.

**Comment by bradenmacdonald** [comment]
Merging now because I need to get this onto our CI machine using its `master` branch. However, this PR can still be used for post-merge code review as needed.

**Pull request merged**